reference/matrix/csr_kernels.cpp

```diff
@@ -71,63 +73,120 @@ namespace reference {
 namespace csr {
 
 
-template <typename ValueType, typename IndexType>
+template <typename MatrixValueType, typename InputValueType,
+          typename OutputValueType, typename IndexType>
 void spmv(std::shared_ptr<const ReferenceExecutor> exec,
-          const matrix::Csr<ValueType, IndexType>* a,
-          const matrix::Dense<ValueType>* b, matrix::Dense<ValueType>* c)
+          const matrix::Csr<MatrixValueType, IndexType>* a,
+          const matrix::Dense<InputValueType>* b,
+          matrix::Dense<OutputValueType>* c)
 {
+    using arithmetic_type =
+        highest_precision<MatrixValueType, InputValueType, OutputValueType>;
+    using a_accessor =
+        gko::acc::reduced_row_major<1, arithmetic_type, const MatrixValueType>;
+    using b_accessor =
+        gko::acc::reduced_row_major<2, arithmetic_type, const InputValueType>;
+    using c_accessor =
+        gko::acc::reduced_row_major<2, arithmetic_type, OutputValueType>;
+
     auto row_ptrs = a->get_const_row_ptrs();
     auto col_idxs = a->get_const_col_idxs();
-    auto vals = a->get_const_values();
+
+    const auto a_vals = gko::acc::range<a_accessor>(
+        std::array<acc::size_type, 1>{
+            static_cast<acc::size_type>(a->get_num_stored_elements())},
+        a->get_const_values());
+    const auto b_vals = gko::acc::range<b_accessor>(
+        std::array<acc::size_type, 2>{
+            {static_cast<acc::size_type>(b->get_size()[0]),
+             static_cast<acc::size_type>(b->get_size()[1])}},
+        b->get_const_values(),
+        std::array<acc::size_type, 1>{
+            {static_cast<acc::size_type>(b->get_stride())}});
+    auto c_vals = gko::acc::range<c_accessor>(
+        std::array<acc::size_type, 2>{
+            {static_cast<acc::size_type>(c->get_size()[0]),
+             static_cast<acc::size_type>(c->get_size()[1])}},
+        c->get_values(),
+        std::array<acc::size_type, 1>{
+            {static_cast<acc::size_type>(c->get_stride())}});
 
     for (size_type row = 0; row < a->get_size()[0]; ++row) {
         for (size_type j = 0; j < c->get_size()[1]; ++j) {
-            c->at(row, j) = zero<ValueType>();
+            c_vals(row, j) = zero<arithmetic_type>();
         }
         for (size_type k = row_ptrs[row];
              k < static_cast<size_type>(row_ptrs[row + 1]); ++k) {
-            auto val = vals[k];
+            auto val = a_vals(k);
             auto col = col_idxs[k];
             for (size_type j = 0; j < c->get_size()[1]; ++j) {
-                c->at(row, j) += val * b->at(col, j);
+                c_vals(row, j) += val * b_vals(col, j);
```

Comment:
this is using the wrong precision to accumulate, we should be using `arithmetic_type` instead of `OutputValueType`. We could either make the changes described in #1170 (blocking, unrolling for small block sizes) or just pull out the rhs loop and accumulate into a local variable